libyul/optimiser/RedundantAssignEliminator.cpp

```diff
@@ -215,101 +81,79 @@ void RedundantAssignEliminator::operator()(Leave const&)
 
 void RedundantAssignEliminator::operator()(Block const& _block)
 {
-	set<YulString> outerDeclaredVariables;
-	swap(m_declaredVariables, outerDeclaredVariables);
+	ScopedSaveAndRestore outerDeclaredVariables(m_declaredVariables, {});
 
-	ASTWalker::operator()(_block);
+	RedundantStoreBase::operator()(_block);
 
 	for (auto const& var: m_declaredVariables)
 		finalize(var, State::Unused);
-
-	swap(m_declaredVariables, outerDeclaredVariables);
 }
 
-
-template <class K, class V, class F>
-void joinMap(std::map<K, V>& _a, std::map<K, V>&& _b, F _conflictSolver)
+void RedundantAssignEliminator::visit(Statement const& _statement)
 {
-	// TODO Perhaps it is better to just create a sorted list
-	// and then use insert(begin, end)
+	RedundantStoreBase::visit(_statement);
 
-	auto ita = _a.begin();
-	auto aend = _a.end();
-	auto itb = _b.begin();
-	auto bend = _b.end();
-
-	for (; itb != bend; ++ita)
-	{
-		if (ita == aend)
-			ita = _a.insert(ita, std::move(*itb++));
-		else if (ita->first < itb->first)
-			continue;
-		else if (itb->first < ita->first)
-			ita = _a.insert(ita, std::move(*itb++));
-		else
-		{
-			_conflictSolver(ita->second, std::move(itb->second));
-			++itb;
-		}
-	}
+	if (auto const* assignment = get_if<Assignment>(&_statement))
+		if (assignment->variableNames.size() == 1)
+			// Default-construct it in "Undecided" state if it does not yet exist.
+			m_stores[assignment->variableNames.front().name][&_statement];
```

Comment:
It has always worked like that apparently, but I'm not entirely clear on why only single assignments are relevant here...

Reply:
On the other hand: why exactly did this move, since it still only applies to the `Assignment` case?

Reply:
Explanation about moving is here: https://github.com/ethereum/solidity/pull/11774/files#r688454105

single-assignments are fine because of SSA: If this is a multi-assignment and turns out to be redundant, the actual SSA assignment will be redundant and removed, while this assignment has to stay unless the right-hand-side is side-effect-free. We have to assign the value of the righ-hand-side somewhere, so this will turn into a variable declaration and the variable will not be used at all.

Reply:
Yul needs a ``pop<n>`` :-).

Reply:
And the reason for multi-assignments to generally not be removed by any of this here, is that they are never actually entered in the map in the first place? Might be worth an assert in ``finalize``... but also fine.